### PR TITLE
[TensorIF] Add custom condition and testcases @open sesame 10/29 09:25 

### PR DESCRIPTION
**Pull request by gichan-jang**
Add testcases for custom condition
    
Signed-off-by: gichan-jang <gichan2.jang@samsung.com>

Self evaluation:
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ *]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/nnstreamer/tensor_if/gsttensorif.c [hunk 1]
```diff
@@ -1097,6 +1098,43 @@ gst_tensor_if_calculate_cv (GstTensorIf * tensor_if, GstBuffer * buf,
...
+ */
+gboolean
+tensor_if_custom_register (GstElement * tensorif_h, tensor_if_custom func,
+    void *data)
+{
+  GstTensorIf *tensor_if = (GstTensorIf *) tensorif_h;
+
+  tensor_if->custom = g_new0 (custom_cb_s, 1);
+
+  if (NULL == tensor_if->custom) {
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_if/gsttensorif.c:1113]: This is unnecessary, and the following lines are never reached.

File: gst/nnstreamer/tensor_if/gsttensorif.c [hunk 2]
```diff
@@ -1097,6 +1098,43 @@ gst_tensor_if_calculate_cv (GstTensorIf * tensor_if, GstBuffer * buf,
...
+
+/**
+ * @brief Unregisters a callback for tensor_if custom condition
+ */
+void
+tensor_if_custom_unregister (GstElement * tensorif_h)
+{
+  GstTensorIf *tensor_if = (GstTensorIf *) tensorif_h;
+
+  if (tensor_if->custom->data) {
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_if/gsttensorif.c:1132]: Unncessary

File: tests/nnstreamer_if/unittest_if.cc [hunk 3]
```diff
@@ -648,6 +648,98 @@ TEST (tensor_if_appsrc, action_0)
...
+  gint *output, i;
+
+  output = (gint *) input->data;
+
+  for (i = 0; i < 48; i++) {
+    if (test_frames[0][i] != output[i]) {
+      *result = FALSE;
+    }
+  }
+  *result = TRUE;
```
Comment (dongju-chae) [inline tests/nnstreamer_if/unittest_if.cc:665]: It seems that `result` always goes to `TRUE`.

File: tests/nnstreamer_if/unittest_if.cc [hunk 4]
```diff
@@ -648,6 +648,98 @@ TEST (tensor_if_appsrc, action_0)
...
+ */
+gboolean tensor_if_custom_cb (const GstTensorIf * tensor_if,
+    const GstTensorMemory * input, gboolean * result) {
+  gint *output, i;
+
+  output = (gint *) input->data;
+
+  for (i = 0; i < 48; i++) {
+    if (test_frames[0][i] != output[i]) {
+      *result = FALSE;
```
Comment (again4you) [inline tests/nnstreamer_if/unittest_if.cc:662]: Exit if failed.

File: gst/nnstreamer/tensor_if/gsttensorif.c (hunk 1 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.c]: Right. Updated

File: gst/nnstreamer/tensor_if/gsttensorif.c (hunk 2 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.c]: Thank you. Updated.

File: tests/nnstreamer_if/unittest_if.cc (hunk 3 above)
Comment (gichan-jang) [reply tests/nnstreamer_if/unittest_if.cc:668]: It's intended.

File: tests/nnstreamer_if/unittest_if.cc (hunk 4 above)
Comment (gichan-jang) [reply tests/nnstreamer_if/unittest_if.cc:666]: Thank you. Updated.

**Review by dongju-chae — APPROVED** [review]
LGTM

File: gst/nnstreamer/tensor_if/gsttensorif.h [hunk 5]
```diff
@@ -140,6 +166,8 @@ struct _GstTensorIf
   GList *cv_option;
   GList *then_option;
   GList *else_option;
+
+  custom_cb_s *custom;
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_if/gsttensorif.h:170]: custom_cb_s has a fixed size. Let's not use pointer here.

File: gst/nnstreamer/tensor_if/gsttensorif.h [hunk 6]
```diff
@@ -35,6 +35,25 @@ G_BEGIN_DECLS
...
+ */
+extern void
+tensor_if_custom_register (GstElement * tensorif_h, tensor_if_custom func, void *data);
+
+/**
+ * @brief Unregister the custom callback function.
+ */
+extern void
+tensor_if_custom_unregister (GstElement * tensorif_h);
+
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_if/gsttensorif.h:56]: These are nnstreamer APIs (not C-API, yet, though)..

We will need an exportable header for these functions.

File: gst/nnstreamer/tensor_if/gsttensorif.h (hunk 5 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.h]: Thank you. Updated.

File: gst/nnstreamer/tensor_if/gsttensorif.h (hunk 6 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.h]: Thank you for the guide.
I uploaded `nnstreamer_plugin_api_if.h`

File: gst/nnstreamer/include/nnstreamer_plugin_api_if.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_if.h:1]: Let's name the file: ```gst/nnstreamer/include/tensor_if.h```

File: gst/nnstreamer/include/nnstreamer_plugin_api_if.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * GStreamer/NNStreamer Tensor-IF
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_if.h:3]: For internal API headers, you need to provide more descriptions: how to use this header.

File: gst/nnstreamer/include/nnstreamer_plugin_api_if.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    const GstTensorMemory *input, gboolean *result);
+
+/**
+ * @brief Register the custom callback function.
+ */
+extern void
+nnstreamer_if_custom_register (GstElement *tensorif_h, tensor_if_custom func, void *data);
+
+/**
+ * @brief Unregister the custom callback function.
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_if.h:37]: This is internal API. You need to fully supply doxygen info: all parameters and return values.

File: gst/nnstreamer/include/nnstreamer_plugin_api_if.h [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+ * @author	Gichan Jang <gichan2.jang@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ */
+
+#ifndef __NNS_PLUGIN_API_IF_H__
+#define __NNS_PLUGIN_API_IF_H__
+
+#include <stdint.h>
+#include "tensor_typedef.h"
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_if.h:20]: Revise header includes: include all headers that are used in this .h file and do not include headers not used in this .h file.
There is no one using stdint and you didn't include glib headers while using glib types. (and so on for GST types)

File: gst/nnstreamer/include/nnstreamer_plugin_api_if.h (hunk 7 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api_if.h]: Changed!

File: gst/nnstreamer/include/nnstreamer_plugin_api_if.h (hunk 8 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api_if.h]: Description is added.  Thanks.

File: gst/nnstreamer/include/nnstreamer_plugin_api_if.h (hunk 10 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api_if.h]: Update header files. Thanks.

File: gst/nnstreamer/include/tensor_if.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * GStreamer/NNStreamer Tensor-IF
+ * Copyright (C) 2020 Gichan Jang <gichan2.jang@samsung.com>
+ */
+/**
+ * @file	nnstreamer_plugin_api_if.h
```
Comment (helloahn) [inline gst/nnstreamer/include/tensor_if.h:7]: incorrect file name

File: gst/nnstreamer/include/tensor_if.h [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    const GstTensorMemory *input, gboolean *result);
+
+/**
+ * @brief Register the custom callback function.
+ * @param[in] tensorif_h tensor_if element from the pipeline
+ * @param[in] func The custom condition function body
+ * @param[in/out] data The internal data for the function
+ */
+extern void
+nnstreamer_if_custom_register (GstElement *tensorif_h, tensor_if_custom func, void *data);
```
Comment (helloahn) [inline gst/nnstreamer/include/tensor_if.h:48]: IMO, there are so many name of `tensor if`, like `nnstreamer_if`, `tensorif`, `tensor_if`. How about making a unified name?

File: gst/nnstreamer/tensor_if/gsttensorif.h [hunk 13]
```diff
@@ -49,6 +49,7 @@ typedef enum {
 				     a specific tensor */
   TIFCV_ALL_TENSORS_AVERAGE_VALUE = 4,	/**< Decide based on a average value of
 					     tensors or a specific tensor */
+  TIFCV_CUSTOM = 5,    /**< Decide based on a user defined condtion */
```
Comment (helloahn) [inline gst/nnstreamer/tensor_if/gsttensorif.h:52]:
```suggestion
  TIFCV_CUSTOM = 5,    /**< Decide based on a user defined condtion */
  TIFCV_CUSTOM = 5,    /**< Decide based on a user defined condition */
```

File: gst/nnstreamer/include/tensor_if.h (hunk 11 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:7]: Thanks! changed

File: gst/nnstreamer/include/tensor_if.h (hunk 12 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:48]: I rearranged the name of the `tensor_if`. Thank you!

File: gst/nnstreamer/tensor_if/gsttensorif.h (hunk 13 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.h:52]: OMG, Thanks!

File: gst/nnstreamer/include/tensor_if.h [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <glib.h>
+#include <gst/gst.h>
+#include "tensor_typedef.h"
+
+G_BEGIN_DECLS
+/**
+ * @brief Calls the user defined conditions function.
+ * @param[in] info Input tensors info.
+ * @param[in] input The array of input tensor data.
```
Comment (myungjoo) [inline gst/nnstreamer/include/tensor_if.h:33]: Misleading. This is not an "array". This is the tensor data itself.

File: gst/nnstreamer/include/tensor_if.h [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+ * @param[out] result result of the user defined condition.
+ * @note The maximum array size of input data is NNS_TENSOR_SIZE_LIMIT.
+ * @return TRUE if there is no error.
+ */
+typedef gboolean (* tensor_if_custom) (const GstTensorsInfo *info,
+    const GstTensorMemory *input, gboolean *result);
+
+/**
+ * @brief Register the custom callback function.
+ * @param[in] tensorif_h tensor_if element from the pipeline
```
Comment (myungjoo) [inline gst/nnstreamer/include/tensor_if.h:43]: What's tensorif_h? How do you get tensorif_h?
Users (tensor-if developers) should be able to understand how to use it and each entity in this header only by reading the doxygen entries.

File: gst/nnstreamer/include/tensor_if.h [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+ * @see		https://github.com/nnstreamer/nnstreamer
+ * @author	Gichan Jang <gichan2.jang@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ * How To for NNdevelopers:
+ *
+ * Define the function in the app.
+ * 1. Define struct, "tensor_if_custom", with the functions defined.
+ * 2. Register the struct with "nnstreamer_if_custom_register" API.
+ * 3. Construct the nnstreamer pipeline and execute it in the app.
```
Comment (myungjoo) [inline gst/nnstreamer/include/tensor_if.h:19]: Need more description: a code example on how to define if-condition and how to use it along with tensor_if element.

Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:19]: I added the code example. Thanks!

File: gst/nnstreamer/include/tensor_if.h (hunk 14 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:33]: Changed parameter description. Thanks

File: gst/nnstreamer/include/tensor_if.h (hunk 15 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:43]: I added how to get the gst element with some reference code.

**Review by gichan-jang — COMMENTED** [review]
I updated the description with some example code. Please review it again.

File: tests/nnstreamer_if/unittest_if.cc [hunk 17]
```diff
@@ -691,6 +691,102 @@ TEST (tensor_if_appsrc, action_1)
...
+
+  sink_handle = gst_bin_get_by_name (GST_BIN (pipeline), "sink_false");
+  EXPECT_NE (sink_handle, nullptr);
+
+  g_signal_connect (sink_handle, "new-data", (GCallback) new_data_cb, (gpointer) &idx);
+  gst_object_unref (sink_handle);
+
+  nnstreamer_if_custom_register (tif_handle, tensor_if_custom_cb, NULL);
+
+  buf_0 = gst_buffer_new ();
```
Comment (helloahn) [inline tests/nnstreamer_if/unittest_if.cc:757]: Is there no problem without unref like `gst_buffer_unref (buf_0);` at the end of this test case?

File: tests/nnstreamer_if/unittest_if.cc [hunk 18]
```diff
@@ -691,6 +691,102 @@ TEST (tensor_if_appsrc, action_1)
...
+  for (i = 0; i < 2; i++) {
+    mem = gst_allocator_alloc (NULL, 192, NULL);
+    gst_memory_map (mem, &info, GST_MAP_WRITE);
+    memcpy (info.data, test_frames[i], 192);
+    gst_memory_unmap (mem, &info);
+    gst_buffer_append_memory (buf_0, mem);
+  }
+  buf_1 = gst_buffer_copy (buf_0);
```
Comment (helloahn) [inline tests/nnstreamer_if/unittest_if.cc:765]: ditto

File: tests/nnstreamer_if/unittest_if.cc (hunk 17 above)
Comment (gichan-jang) [reply tests/nnstreamer_if/unittest_if.cc:757]: Yes. `gst_app_src_push_buffer` takes ownership of the buffer. So don't need to unref buffer.

File: gst/nnstreamer/tensor_if/gsttensorif.c [hunk 19]
```diff
@@ -166,6 +166,7 @@ gst_tensor_if_cv_get_type (void)
       {TIFCV_A_VALUE, "A_VALUE", "Decide based on a single scalar value"},
       {TIFCV_TENSOR_AVERAGE_VALUE, "TENSOR_AVERAGE_VALUE",
           "Decide based on a average value of a specific tensor"},
+      {TIFCV_CUSTOM, "CUSTOM", "Decide based on a user defined condtion"},
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_if/gsttensorif.c:169]: I think it is good to flesh out the word condition.
```suggestion
      {TIFCV_CUSTOM, "CUSTOM", "Decide based on a user defined callback"},
```

File: gst/nnstreamer/tensor_if/gsttensorif.c [hunk 20]
```diff
@@ -1097,6 +1099,31 @@ gst_tensor_if_calculate_cv (GstTensorIf * tensor_if, GstBuffer * buf,
...
 }
 
+/**
+ * @brief Registers a callback for tensor_if custom condition
+ */
+void
+nnstreamer_if_custom_register (GstElement * element, tensor_if_custom func,
+    void *data)
+{
+  GstTensorIf *tensor_if = GST_TENSOR_IF (element);
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_if/gsttensorif.c:1109]: Q. is it okay if `func == NULL`?

File: gst/nnstreamer/tensor_if/gsttensorif.c [hunk 21]
```diff
@@ -287,6 +288,7 @@ gst_tensor_if_init (GstTensorIf * tensor_if)
 
   tensor_if->num_srcpads = 0;
   tensor_if->srcpads = NULL;
+  tensor_if->custom.func = NULL;
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_if/gsttensorif.c:291]: `custom.data = NULL`  as well?

File: gst/nnstreamer/tensor_if/gsttensorif.c [hunk 22]
```diff
@@ -1107,13 +1134,36 @@ static gboolean
...
+      }
+      in_tensors[i].data = in_info[i].data;
+      in_tensors[i].size = in_info[i].size;
+      gst_memory_unmap (in_mem[i], &in_info[i]);
+    }
 
-  if (!gst_tensor_if_calculate_cv (tensor_if, buf, &cv)) {
-    GST_ERROR_OBJECT (tensor_if, " failed to calculate compared value");
-    return FALSE;
+    return tensor_if->custom.func (&tensor_if->in_config.info, in_tensors,
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_if/gsttensorif.c:1157]: Could you add a test with user custom data? I think passing user-defined data is omitted here

File: gst/nnstreamer/tensor_if/gsttensorif.c (hunk 19 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.c:169]: Updated.

File: gst/nnstreamer/tensor_if/gsttensorif.c (hunk 21 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.c:291]: Added! Thanks.

File: gst/nnstreamer/tensor_if/gsttensorif.c (hunk 20 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.c:1109]: Added! Thanks.

File: gst/nnstreamer/tensor_if/gsttensorif.c (hunk 22 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.c:1157]: I added the case in  `TEST (tensor_if_custom, normal_0)`. Thanks.

**Review by zhoonit — APPROVED** [review]
LGTM :)

File: gst/nnstreamer/include/tensor_if.h [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+ *   g_usleep (100000);
+ *   gst_element_set_state (pipeline, GST_STATE_NULL);
+ *   g_usleep (100000);
+ *
+ *   nnstreamer_if_custom_unregister (tif_handle);
+ *   gst_object_unref (tif_handle);
+ *   gst_object_unref (pipeline);
+ *   g_free (str_pipeline);
+ * }
+ * @endcode
```
Comment (myungjoo) [inline gst/nnstreamer/include/tensor_if.h:61]: You don't need the full source to run. You just need to show how to use custom IF condition:
```
 * Usage example of the tensor_if custom condition
 * @code
 * // Define custom callback function describing the condition
 * gboolean tensor_if_custom_cb (const GstTensorsInfo *info,
 *     const GstTensorMemory * input, gboolean * result) {
 *   // Describe the conditions and pass the results
 * }
 *
 * ...
 * 
 *   // Get the handle of the tensor_if element
 *   tif_handle = gst_bin_get_by_name (GST_BIN (pipeline), "tif");
 *   // Register custom callback function describing the condition
 *   nnstreamer_if_custom_register (tif_handle, tensor_if_custom_cb, NULL);
 *
 * ...
 *   // Use the condition in a pipeline
 *   // E.g., Pipeline of " ... ! tensor_if name=tif compared_value=CUSTOM then=TENSORPICK then-option=0 else=TENSORPICK else-option=1 ! ... "
 *
 * ...
 *   // After everything is done.
 *   nnstreamer_if_custom_unregister (tif_handle);
 *   gst_object_unref (tif_handle);
 * @endcode
```

Comment (myungjoo) [reply gst/nnstreamer/include/tensor_if.h:61]: It'd be more nice if you can do:
```
nnstreamer_if_custom_register ("if_X", tensor_if_custom_cb, NULL);
...
PIPELINE = " ... ! tensor_if compared_value=CUSTOM,if_X ... ! ... "
...
nnstreamer_if_custom_unregister("if_X");
```

Comment (myungjoo) [reply gst/nnstreamer/include/tensor_if.h:61]: glib has hash-table. You won't need to reimplement the data structure, too.

Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:44]: Changed to register custom callback by the name not element.
for example
```
nnstreamer_if_custom_register ("if_X", tensor_if_custom_cb, NULL);
...
PIPELINE = " ... ! tensor_if compared-value=CUSTOM compared-value-option=if_X ... ! ... "
...
nnstreamer_if_custom_unregister("if_X");
```

Comment (myungjoo) [reply gst/nnstreamer/include/tensor_if.h:44]: Uh.. you need to register "tifx" before using the pipeline. :)

File: gst/nnstreamer/nnstreamer_subplugin.c [hunk 24]
```diff
@@ -68,6 +68,7 @@ static subpluginSearchLogic searchAlgorithm[] = {
   [NNS_SUBPLUGIN_DECODER] = NNS_SEARCH_FILENAME,
   [NNS_EASY_CUSTOM_FILTER] = NNS_SEARCH_FILENAME,
   [NNS_SUBPLUGIN_CONVERTER] = NNS_SEARCH_GETALL,
+  [NNS_IF_CUSTOM] = NNS_SEARCH_FILENAME,
```
Comment (myungjoo) [inline gst/nnstreamer/nnstreamer_subplugin.c:71]: NNS_SEARCH_NO_OP? unless you intend to enable "nnstreamer_customif_whatever.so" to be loaded.

File: gst/nnstreamer/tensor_if/gsttensorif.c [hunk 25]
```diff
@@ -1107,13 +1190,41 @@ static gboolean
...
-  tensor_if_data_s cv = {.type = _NNS_END,.data._uint8_t = 0 };
+  if (tensor_if->cv == TIFCV_CUSTOM) {
+    GstMemory *in_mem[NNS_TENSOR_SIZE_LIMIT];
+    GstMapInfo in_info[NNS_TENSOR_SIZE_LIMIT];
+    GstTensorMemory in_tensors[NNS_TENSOR_SIZE_LIMIT];
+    guint i, j;
+
+    if (tensor_if->custom.func == NULL) {
+      const custom_cb_s *ptr =
+          get_subplugin (NNS_IF_CUSTOM, tensor_if->custom_name);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_if/gsttensorif.c:1201]: get_subplugin is not that cheap. you may need to cache it (reload if set_property is called or related events occur)

File: gst/nnstreamer/include/tensor_if.h (hunk 23 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:44]: Update the comment. Thanks.

Comment (gichan-jang) [reply gst/nnstreamer/include/tensor_if.h:44]: Right! Updated.

File: gst/nnstreamer/nnstreamer_subplugin.c (hunk 24 above)
Comment (gichan-jang) [reply gst/nnstreamer/nnstreamer_subplugin.c]: Thanks. Changed to `NNS_SEARCH_NO_OP`

File: gst/nnstreamer/tensor_if/gsttensorif.c (hunk 25 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_if/gsttensorif.c]: Changed to set custom properties in set_property.
And if a user tries to set a new custom option, simply overwrite it.

**Review by myungjoo — APPROVED** [review]
LGTM

**Pull request merged**